### PR TITLE
Add forgot password flow

### LINKED ISSUE #244: Forgot Password Flow
**Goal**
As an applicant, I want to be able create a new password so that I can access my account

**Acceptance Criteria**
- There should be a forgot password link available on the Sign In screen alongside the password field
- Clicking on this link should take the applicant to screen which prompts them for their email address in order to send them a link
- The email used in this field should be associated with an existing user account
- If the email entered does not match an existing user account the applicant should receive a validation warning
- Email not found. Please make sure your email has an account with us and is confirmed.
- If a valid email address is entered the applicant should see a notification
   _**We've sent you an email.**
   You'll receive an email with a link to reset your password._
- The email the applicant received should give them a message prompting them to change  their password:
     _A request to reset your Bloom Housing Portal website password for (web url) has recently been made.
     If you didn't request this, please ignore this email.
     If you did make this request, please click on the link below to reset your password:
     [Change my password](url)
     Your password won't change until you access the link above and create a new one_
- The applicant should be redirected to a page which allows the applicant to enter their new password twice for confirmation.
- If the applicant does not enter the same password in both field an inline error messages states: 
   _Password confirmation does not match_
- If the applicant does enter the same pass both times they are redirected to the My Applications page of Accounts

**Documentation**
[Dahlia Forgot Password](https://dahlia-full.herokuapp.com/forgot-password)
image

image

**Engineering Notes**

### PR DESCRIPTION
**Pull request by software-project**
Resolves #244 

I need to write some test, but should be ready for a review.

### CODE REVIEW

**Comment by jaredcwhite** [comment]
@software-project Approach seems sound to me. One thought: should we have 24 hours for the reset token expiration? I've generally seen short timeframes, as low as 15 minutes even. (I remember for one service it was actually a problem because I was experiencing a delay in mail delivery and by the time I'd get the reset message, the link had already expired!) Devise looks like it defaults to 6 hours, FYI.

**Review by pbn4 — APPROVED** [review]
Good stuff. :)

File: backend/core/src/shared/email.service.ts [hunk 1]
```diff
@@ -90,6 +90,29 @@ export class EmailService {
     )
   }
 
+  public async forgotPassword(user: User, appUrl: string) {
+    const compiledTemplate = this.template("forgot-password")
+    const resetUrl = `${appUrl}/reset-password?token=${user.resetToken}`
+
+    if (this.configService.get<string>("NODE_ENV") == "production") {
```
[the remark this answers is not in the document]
Comment (software-project) [reply backend/core/src/shared/email.service.ts:97]: Thats something Ben put in the previous email service, so I copied it over.

File: backend/core/src/user/dto/forgot_password.dto.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+export class ForgotPasswordDto {
+  @Expose()
+  @IsEmail({}, { groups: [ValidationsGroupsEnum.default] })
+  @MaxLength(256, { groups: [ValidationsGroupsEnum.default] })
+  email: string
+
+  @Expose()
+  @IsOptional({ groups: [ValidationsGroupsEnum.default] })
+  @IsString({ groups: [ValidationsGroupsEnum.default] })
+  @MaxLength(256, { groups: [ValidationsGroupsEnum.default] })
```
Comment (pbn4) [inline backend/core/src/user/dto/forgot_password.dto.ts:14]: @software-project Is there a chance that this will be too small for an actual URL?

**Review by pbn4 — CHANGES_REQUESTED** [review]
LGTM

File: backend/core/src/user/dto/forgot_password.dto.ts (hunk 2 above)
Comment (software-project) [reply backend/core/src/user/dto/forgot_password.dto.ts:14]: @pbn4 I don't think so. It's just the base app url.

**Review by pbn4 — APPROVED** [review]

**Pull request merged**